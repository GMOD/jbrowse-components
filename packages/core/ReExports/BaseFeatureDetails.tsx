import { lazy } from 'react'

import { lazyMap } from './lazify'

export const BaseFeatureDetail = lazyMap(
  {
    Attributes: lazy(
      () => import('../BaseFeatureWidget/BaseFeatureDetail/Attributes'),
    ),
    FeatureDetails: lazy(
      () => import('../BaseFeatureWidget/BaseFeatureDetail/FeatureDetails'),
    ),
    BaseCard: lazy(
      () => import('../BaseFeatureWidget/BaseFeatureDetail/BaseCard'),
    ),
    BaseAttributes: lazy(
      () => import('../BaseFeatureWidget/BaseFeatureDetail/BaseAttributes'),
    ),
    BaseCoreDetails: lazy(
      () => import('../BaseFeatureWidget/BaseFeatureDetail/BaseCoreDetails'),
    ),
  },
  '@jbrowse/core/BaseFeatureWidget/BaseFeatureDetail/',
)
