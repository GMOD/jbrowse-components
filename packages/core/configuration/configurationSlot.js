import { types } from 'mobx-state-tree'
import { stringToFunction, functionRegexp } from '../util/functionStrings'
import { inDevelopment } from '../util'
import { FileLocation } from '../mst-types'

function isValidColorString(/* str */) {
  // TODO: check all the crazy cases for whether it's a valid HTML/CSS color string
  return true
}
const typeModels = {
  stringArray: types.array(types.string),
  stringArrayMap: types.map(types.array(types.string)),
  numberMap: types.map(types.number),
  boolean: types.boolean,
  color: types.refinement('Color', types.string, isValidColorString),
  integer: types.integer,
  number: types.number,
  string: types.string,
  text: types.string,
  fileLocation: FileLocation,
  frozen: types.frozen(),
}

// default values we use if the defaultValue is malformed or does not work
const fallbackDefaults = {
  stringArray: [],
  stringArrayMap: {},
  numberMap: {},
  boolean: true,
  color: 'black',
  integer: 1,
  number: 1,
  string: '',
  text: '',
  fileLocation: { uri: '/path/to/resource.txt' },
  frozen: {},
}

const literalJSON = self => ({
  views: {
    get valueJSON() {
      return self.value
    },
  },
})

const objectJSON = self => ({
  views: {
    get valueJSON() {
      return JSON.stringify(self.value)
    },
  },
})

// custom actions for modifying the value models
const typeModelExtensions = {
  fileLocation: objectJSON,
  number: literalJSON,
  integer: literalJSON,
  boolean: literalJSON,
  frozen: objectJSON,
  // special actions for working with stringArray slots
  stringArray: self => ({
    views: {
      get valueJSON() {
        return JSON.stringify(self.value)
      },
    },
    actions: {
      add(val) {
        self.value.push(val)
      },
      removeAtIndex(idx) {
        self.value.splice(idx, 1)
      },
      setAtIndex(idx, val) {
        self.value[idx] = val
      },
    },
  }),
  stringArrayMap: self => ({
    views: {
      get valueJSON() {
        return JSON.stringify(self.value)
      },
    },
    actions: {
      add(key, val) {
        self.value.set(key, val)
      },
      remove(key) {
        self.value.delete(key)
      },
      addToKey(key, val) {
        self.value.get(key).push(val)
      },
      removeAtKeyIndex(key, idx) {
        self.value.get(key).splice(idx, 1)
      },
      setAtKeyIndex(key, idx, val) {
        self.value.get(key)[idx] = val
      },
    },
  }),
  numberMap: self => ({
    views: {
      get valueJSON() {
        return JSON.stringify(self.value)
      },
    },
    actions: {
      add(key, val) {
        self.value.set(key, val)
      },
      remove(key) {
        self.value.delete(key)
      },
    },
  }),
}

const FunctionStringType = types.refinement(
  'FunctionString',
  types.string,
  str => functionRegexp.test(str),
)

/**
 * builds a MST model for a configuration slot
 *
 * @param {*} slotName
 * @param {object} definition
 * @param {string} definition.description?
 * @param {object} definition.model? custom base MST model for the slot's value
 * @param {string} definition.type name of the type of slot, e.g. "string", "number", "stringArray"
 * @param {*} definition.defaultValue default value of the slot
 * @param {Array[string]} [definition.functionSignature] parameter names
 *  of the function callback, default []
 */
export default function ConfigSlot(
  slotName,
  { description = '', model, type, defaultValue, functionSignature = [] },
) {
  if (!type) throw new Error('type name required')
  if (!model) model = typeModels[type]
  if (!model) {
    throw new Error(
      `no builtin config slot type "${type}", and no 'model' param provided`,
    )
  }

  if (defaultValue === undefined) throw new Error("no 'defaultValue' provided")

  // if the `type` is something like `color`, then the model name
  // here will be `ColorConfigSlot`
  const configSlotModelName = `${slotName
    .charAt(0)
    .toUpperCase()}${slotName.slice(1)}ConfigSlot`
  let slot = types
    .model(configSlotModelName, {
      name: types.literal(slotName),
      description: types.literal(description),
      type: types.literal(type),
      value: types.optional(
        types.union(FunctionStringType, model),
        defaultValue,
      ),
    })
    .volatile(() => ({
      functionSignature,
    }))
    .views(self => ({
      get func() {
        if (self.isCallback) {
          // compile this as a function
          return stringToFunction(String(self.value), {
            verifyFunctionSignature: inDevelopment
              ? functionSignature
              : undefined,
          })
        }
        return () => self.value
      },
      get isCallback() {
        return /^\s*function\s*\(/.test(self.value)
      },

      // JS representation of the value of this slot, suitable
      // for embedding in either JSON or a JS function string.
      // many of the data types override this in typeModelExtensions
      get valueJSON() {
        if (self.isCallback) return undefined
        return self.value && self.value.toJSON
          ? self.value.toJSON()
          : `'${self.value}'`
      },
    }))
    .preProcessSnapshot(
      val =>
        typeof val === 'object' && val.name === slotName
          ? val
          : {
              name: slotName,
              description,
              type,
              value: val,
            },
      // ({
      //   name: slotName,
      //   description,
      //   type,
      //   value: val,
      // }),
    )
    .postProcessSnapshot(snap => {
      if (typeof snap.value === 'object')
        return JSON.stringify(snap.value) !== JSON.stringify(defaultValue)
          ? snap.value
          : undefined
      return snap.value !== defaultValue ? snap.value : undefined
    })
    .actions(self => ({
      set(newVal) {
        self.value = newVal
      },
      reset() {
        self.value = defaultValue
      },
      convertToCallback() {
        if (self.isCallback) return
        self.value = `function(${self.functionSignature.join(', ')}) {
  return ${self.valueJSON}
}
`
      },
      convertToValue() {
        if (!self.isCallback) return
        // try calling it with no arguments
        try {
          const funcResult = self.func()
          if (funcResult !== undefined) {
            self.value = funcResult
            return
          }
        } catch (e) {
          /* ignore */
        }
        self.value = defaultValue
        // if it is still a callback (happens if the defaultValue is a callback),
        // then use the last-resort fallback default
        if (self.isCallback) {
          if (!(type in fallbackDefaults)) {
            throw new Error(`no fallbackDefault defined for type ${type}`)
          }
          self.value = fallbackDefaults[type]
        }
      },
    }))

  // if there are any type-specific extensions (views or actions)
  //  to the slot, add those in
  if (typeModelExtensions[type]) {
    slot = slot.extend(typeModelExtensions[type])
  }

  const completeModel = types.optional(slot, {
    name: slotName,
    type,
    description,
    value: defaultValue,
  })
  completeModel.isJBrowseConfigurationSlot = true
  return completeModel
}
