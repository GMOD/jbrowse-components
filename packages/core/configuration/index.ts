export {
  ConfigurationSchema,
  ConfigurationReference,
} from './configurationSchema'

export type {
  AnyConfigurationSchemaType,
  AnyConfigurationModel,
  AnyConfigurationSlot,
  AnyConfigurationSlotType,
  AnyConfiguration,
} from './types'

export * from './util'
