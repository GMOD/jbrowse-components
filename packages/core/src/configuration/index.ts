export {
  ConfigurationSchema,
  ConfigurationReference,
} from './configurationSchema'

export * from './util'
