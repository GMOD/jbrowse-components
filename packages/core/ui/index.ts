export * from './theme'
export { LogoFull, Logomark } from './Logo'
export { default as App } from './App'
export { default as FileSelector } from './FileSelector'
export { default as PrerenderedCanvas } from './PrerenderedCanvas'
export { default as ResizeHandle } from './ResizeHandle'
export { default as StartScreen } from './StartScreen'
export { default as EditableTypography } from './EditableTypography'
export { default as FactoryResetDialog } from './FactoryResetDialog'
export { default as Tooltip } from './Tooltip'
export { default as FatalErrorDialog } from './FatalErrorDialog'
export { default as Menu } from './Menu'
export { default as SanitizedHTML } from './SanitizedHTML'
export * from './Menu'
