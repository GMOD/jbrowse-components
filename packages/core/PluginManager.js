import React from 'react'
import * as mst from 'mobx-state-tree'

import RendererType from './pluggableElementTypes/renderers/RendererType'
import AdapterType from './pluggableElementTypes/AdapterType'
import TrackType from './pluggableElementTypes/TrackType'
import ViewType from './pluggableElementTypes/ViewType'
import DrawerWidgetType from './pluggableElementTypes/DrawerWidgetType'
import MenuBarType from './pluggableElementTypes/MenuBarType'
import ConnectionType from './pluggableElementTypes/ConnectionType'

import { ConfigurationSchema } from './configuration'

// little helper class that keeps groups of callbacks that are
// then run in a specified order by group
class PhasedScheduler {
  phaseCallbacks = {}

  constructor(...phaseOrder) {
    this.phaseOrder = phaseOrder
  }

  add(phase, callback) {
    if (!this.phaseOrder.includes(phase)) {
      throw new Error(`unknown phase ${phase}`)
    }
    if (!this.phaseCallbacks[phase]) this.phaseCallbacks[phase] = []
    this.phaseCallbacks[phase].push(callback)
  }

  run() {
    this.phaseOrder.forEach(phaseName => {
      if (this.phaseCallbacks[phaseName]) {
        this.phaseCallbacks[phaseName].forEach(callback => callback())
      }
    })
  }
}

export default class PluginManager {
  plugins = []

  elementTypes = {}

  elementCreationSchedule = new PhasedScheduler(
    'renderer',
    'adapter',
    'track',
    'connection',
    'view',
    'drawer widget',
    'menu bar',
  )

  typeBaseClasses = {
    renderer: RendererType,
    adapter: AdapterType,
    track: TrackType,
    connection: ConnectionType,
    view: ViewType,
    'drawer widget': DrawerWidgetType,
    'menu bar': MenuBarType,
  }

  static lib = { 'mobx-state-tree': mst, React }

  constructor(initialPlugins = []) {
    this.lib = PluginManager.lib

    this.getRendererType = this.getElementType.bind(this, 'renderer')
    this.getAdapterType = this.getElementType.bind(this, 'adapter')
    this.getTrackType = this.getElementType.bind(this, 'track')
    this.getViewType = this.getElementType.bind(this, 'view')
    this.getDrawerWidgetType = this.getElementType.bind(this, 'drawer widget')
    this.getMenuBarType = this.getElementType.bind(this, 'menu bar')
    this.getConnectionType = this.getElementType.bind(this, 'connection')
    this.addRendererType = this.addElementType.bind(this, 'renderer')
    this.addAdapterType = this.addElementType.bind(this, 'adapter')
    this.addTrackType = this.addElementType.bind(this, 'track')
    this.addViewType = this.addElementType.bind(this, 'view')
    this.addDrawerWidgetType = this.addElementType.bind(this, 'drawer widget')
    this.addMenuBarType = this.addElementType.bind(this, 'menu bar')
    this.addConnectionType = this.addElementType.bind(this, 'connection')

    // add all the initial plugins
    initialPlugins.forEach(plugin => {
      this.addPlugin(plugin)
    })
  }

  addPlugin(plugin) {
    if (this.configured) {
      throw new Error('JBrowse already configured, cannot add plugins')
    }
    if (this.plugins.includes(plugin)) {
      throw new Error('plugin already installed')
    }
    // if (!plugin.install) console.error(plugin)
    plugin.install(this)
    this.plugins.push(plugin)
    return this
  }

  /** get a MST type for the union of all specified pluggable MST types */
  pluggableMstType(
    typeGroup,
    fieldName,
    fallback = mst.types.maybe(mst.types.null),
  ) {
    const pluggableTypes = this.getElementTypeMembers(typeGroup, fieldName)
    // try to smooth over the case when no types are registered, mostly encountered in tests
    if (pluggableTypes.length === 0) return fallback
    return mst.types.union(...pluggableTypes)
  }

  /** get a MST type for the union of all specified pluggable config schemas */
  pluggableConfigSchemaType(
    typeGroup,
    fieldName = 'configSchema',
    fallback = ConfigurationSchema('Null', {}),
  ) {
    return this.pluggableMstType(typeGroup, fieldName, fallback)
  }

  addElementType(groupName, creationCallback) {
    if (typeof creationCallback !== 'function') {
      throw new Error('must provide a callback function')
    }
    const typeBaseClass = this.typeBaseClasses[groupName]
    if (!typeBaseClass) {
      throw new Error(`unknown pluggable element type ${groupName}, cannot add`)
    }
    if (!this.elementTypes[groupName]) this.elementTypes[groupName] = {}

    this.elementCreationSchedule.add(groupName, () => {
      const element = creationCallback(this)
      if (
        groupName === 'adapter' &&
        !element.AdapterClass.capabilities.length
      ) {
        throw new Error(
          `Adapter ${
            element.AdapterClass.name
          } must provide a static property "capabilities" that has at least one entry. See BaseAdapter for an example.`,
        )
      }
      if (this.elementTypes[groupName][element.name]) {
        throw new Error(
          `${groupName} ${
            element.name
          } already registered, cannot register it again`,
        )
      }

      this.elementTypes[groupName][element.name] = element
    })

    return this
  }

  getElementType(groupName, typeName) {
    if (!typeName) {
      throw new Error(`invalid type name "${typeName}"`)
    }
    return (this.elementTypes[groupName] || {})[typeName]
  }

  getElementTypesInGroup(groupName) {
    return Object.values(this.elementTypes[groupName] || {})
  }

  getElementTypeMembers(groupName, memberName) {
    return this.getElementTypesInGroup(groupName)
      .map(t => t[memberName])
      .filter(m => !!m)
  }

  configure() {
    if (this.configured) throw new Error('already configured')

    // run the creation callbacks for each element type in order.
    // see elementCreationSchedule above for the creation order
    this.elementCreationSchedule.run()
    delete this.elementCreationSchedule

    this.plugins.forEach(plugin => plugin.configure())

    this.configured = true

    // console.log(JSON.stringify(getSnapshot(model)))

    return this
  }
}
