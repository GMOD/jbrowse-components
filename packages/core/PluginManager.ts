import {
  types,
  IAnyType,
  IAnyModelType,
  isModelType,
  isType,
} from 'mobx-state-tree'

import PluggableElementBase from './pluggableElementTypes/PluggableElementBase'
import RendererType from './pluggableElementTypes/renderers/RendererType'
import AdapterType from './pluggableElementTypes/AdapterType'
import TrackType from './pluggableElementTypes/TrackType'
import DisplayType from './pluggableElementTypes/DisplayType'
import ViewType from './pluggableElementTypes/ViewType'
import WidgetType from './pluggableElementTypes/WidgetType'
import ConnectionType from './pluggableElementTypes/ConnectionType'
import RpcMethodType from './pluggableElementTypes/RpcMethodType'

import {
  ConfigurationSchema,
  isBareConfigurationSchemaType,
} from './configuration'

import Plugin from './Plugin'
import ReExports from './ReExports'

import {
  PluggableElementType,
  PluggableElementMember,
} from './pluggableElementTypes'
import { AnyConfigurationSchemaType } from './configuration/configurationSchema'
import { AbstractRootModel } from './util'
import CorePlugin from './CorePlugin'

/** little helper class that keeps groups of callbacks that are
then run in a specified order by group */
class PhasedScheduler<PhaseName extends string> {
  phaseCallbacks = new Map<PhaseName, Function[]>()

  phaseOrder: PhaseName[] = []

  constructor(...phaseOrder: PhaseName[]) {
    this.phaseOrder = phaseOrder
  }

  add(phase: PhaseName, callback: Function) {
    if (!this.phaseOrder.includes(phase)) {
      throw new Error(`unknown phase ${phase}`)
    }
    let phaseCallbacks = this.phaseCallbacks.get(phase)
    if (!phaseCallbacks) {
      phaseCallbacks = []
      this.phaseCallbacks.set(phase, phaseCallbacks)
    }
    phaseCallbacks.push(callback)
  }

  run() {
    this.phaseOrder.forEach(phaseName => {
      const phaseCallbacks = this.phaseCallbacks.get(phaseName)
      if (phaseCallbacks) {
        phaseCallbacks.forEach(callback => callback())
      }
    })
  }
}

type PluggableElementTypeGroup =
  | 'renderer'
  | 'adapter'
  | 'display'
  | 'track'
  | 'connection'
  | 'view'
  | 'widget'
  | 'rpc method'

/** internal class that holds the info for a certain element type */
class TypeRecord<ElementClass extends PluggableElementBase> {
  registeredTypes: { [name: string]: ElementClass } = {}

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  baseClass: { new (...args: any[]): ElementClass }

  typeName: string

  constructor(
    typeName: string,
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    elementType: { new (...args: any[]): ElementClass },
  ) {
    this.typeName = typeName
    this.baseClass = elementType
  }

  add(name: string, t: ElementClass) {
    this.registeredTypes[name] = t
  }

  has(name: string) {
    return name in this.registeredTypes
  }

  get(name: string) {
    if (!this.has(name))
      throw new Error(
        `${this.typeName} '${name}' not found, perhaps its plugin is not loaded or its plugin has not added it.`,
      )
    return this.registeredTypes[name]
  }

  all() {
    return Object.values(this.registeredTypes)
  }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type AnyFunction = (...args: any) => any

export default class PluginManager {
  plugins: Plugin[] = []

  elementCreationSchedule = new PhasedScheduler<PluggableElementTypeGroup>(
    'renderer',
    'adapter',
    'display',
    'track',
    'connection',
    'view',
    'widget',
    'rpc method',
  )

  rendererTypes = new TypeRecord('RendererType', RendererType)

  adapterTypes = new TypeRecord('AdapterType', AdapterType)

  trackTypes = new TypeRecord('TrackType', TrackType)

  displayTypes = new TypeRecord('DisplayType', DisplayType)

  connectionTypes = new TypeRecord('ConnectionType', ConnectionType)

  viewTypes = new TypeRecord('ViewType', ViewType)

  widgetTypes = new TypeRecord('WidgetType', WidgetType)

  rpcMethods = new TypeRecord('RpcMethodType', RpcMethodType)

  configured = false

  rootModel?: AbstractRootModel

  constructor(initialPlugins: Plugin[] = []) {
    // add the core plugin
    this.addPlugin(new CorePlugin())

    // add all the initial plugins
    initialPlugins.forEach(plugin => {
      this.addPlugin(plugin)
    })
  }

  addPlugin(plugin: Plugin) {
    if (this.configured) {
      throw new Error('JBrowse already configured, cannot add plugins')
    }
    if (this.plugins.includes(plugin)) {
      throw new Error('plugin already installed')
    }
    plugin.install(this)
    this.plugins.push(plugin)
    return this
  }

  getPlugin(name: string) {
    return this.plugins.find(p => p.name === name)
  }

  hasPlugin(name: string) {
    return this.getPlugin(name) !== undefined
  }

  createPluggableElements() {
    // run the creation callbacks for each element type in order.
    // see elementCreationSchedule above for the creation order
    this.elementCreationSchedule.run()
    delete this.elementCreationSchedule
    return this
  }

  setRootModel(rootModel: AbstractRootModel) {
    this.rootModel = rootModel
  }

  configure() {
    if (this.configured) throw new Error('already configured')

    this.plugins.forEach(plugin => plugin.configure(this))

    this.configured = true

    // console.log(JSON.stringify(getSnapshot(model)))

    return this
  }

  getElementTypeRecord(
    groupName: PluggableElementTypeGroup,
  ): TypeRecord<PluggableElementBase> {
    switch (groupName) {
      case 'adapter':
        return this.adapterTypes
      case 'connection':
        return this.connectionTypes
      case 'widget':
        return this.widgetTypes
      case 'renderer':
        return this.rendererTypes
      case 'display':
        return this.displayTypes
      case 'track':
        return this.trackTypes
      case 'view':
        return this.viewTypes
      case 'rpc method':
        return this.rpcMethods
      default:
        throw new Error(`invalid element type '${groupName}'`)
    }
  }

  addElementType(
    groupName: PluggableElementTypeGroup,
    creationCallback: (pluginManager: PluginManager) => PluggableElementType,
  ) {
    if (typeof creationCallback !== 'function') {
      throw new Error(
        'must provide a callback function that returns the new type object',
      )
    }
    const typeRecord = this.getElementTypeRecord(groupName)

    this.elementCreationSchedule.add(groupName, () => {
      const newElement = creationCallback(this)
      if (!newElement.name)
        throw new Error(`cannot add a ${groupName} with no name`)

      if (typeRecord.has(newElement.name)) {
        throw new Error(
          `${groupName} ${newElement.name} already registered, cannot register it again`,
        )
      }

      typeRecord.add(newElement.name, newElement)
    })

    return this
  }

  getElementType(groupName: PluggableElementTypeGroup, typeName: string) {
    const typeRecord = this.getElementTypeRecord(groupName)
    return typeRecord.get(typeName)
  }

  getElementTypesInGroup(groupName: PluggableElementTypeGroup) {
    const typeRecord = this.getElementTypeRecord(groupName)
    return typeRecord.all()
  }

  /** get a MST type for the union of all specified pluggable MST types */
  pluggableMstType(
    typeGroup: PluggableElementTypeGroup,
    fieldName: PluggableElementMember,
    fallback: IAnyType = types.maybe(types.null),
  ) {
    const pluggableTypes: IAnyModelType[] = []
    this.getElementTypeRecord(typeGroup)
      .all()
      .forEach(t => {
        const thing = t[fieldName]
        if (isType(thing) && isModelType(thing)) {
          pluggableTypes.push(thing)
        }
      })
    // try to smooth over the case when no types are registered, mostly encountered in tests
    if (pluggableTypes.length === 0) {
      console.warn(
        `No JBrowse pluggable types found matching ('${typeGroup}','${fieldName}')`,
      )
      return fallback
    }
    return types.union(...pluggableTypes)
  }

  /** get a MST type for the union of all specified pluggable config schemas */
  pluggableConfigSchemaType(
    typeGroup: PluggableElementTypeGroup,
    fieldName: PluggableElementMember = 'configSchema',
  ) {
    const pluggableTypes: AnyConfigurationSchemaType[] = []
    this.getElementTypeRecord(typeGroup)
      .all()
      .forEach(t => {
        const thing = t[fieldName]
        if (isBareConfigurationSchemaType(thing)) {
          pluggableTypes.push(thing)
        }
      })
    if (pluggableTypes.length === 0)
      pluggableTypes.push(ConfigurationSchema('Null', {}))
    return types.union(...pluggableTypes)
  }

  jbrequireCache = new Map()

  lib = ReExports

  load = <FTYPE extends AnyFunction>(lib: FTYPE): ReturnType<FTYPE> => {
    if (!this.jbrequireCache.has(lib)) this.jbrequireCache.set(lib, lib(this))
    return this.jbrequireCache.get(lib)
  }

  /**
   * Get the re-exported version of the given package name.
   * Throws an error if the package is not re-exported by the plugin manager.
   *
   * @returns the library's default export
   */
  jbrequire = (
    lib: keyof typeof ReExports | AnyFunction | { default: AnyFunction },
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ): any => {
    if (typeof lib === 'string') {
      const pack = this.lib[lib]
      if (!pack) {
        throw new TypeError(
          `No jbrequire re-export defined for package '${lib}'. If this package must be shared between plugins, add it to ReExports.js. If it does not need to be shared, just import it normally.`,
        )
      }
      return pack
    }

    if (typeof lib === 'function') {
      return this.load(lib)
    }

    if (lib.default) return this.jbrequire(lib.default)

    throw new TypeError(
      'lib passed to jbrequire must be either a string or a function',
    )
  }

  getRendererType(typeName: string): RendererType {
    return this.rendererTypes.get(typeName)
  }

  getRendererTypes(): RendererType[] {
    return this.rendererTypes.all()
  }

  getAdapterType(typeName: string): AdapterType {
    return this.adapterTypes.get(typeName)
  }

  getTrackType(typeName: string): TrackType {
    return this.trackTypes.get(typeName)
  }

  getDisplayType(typeName: string): DisplayType {
    return this.displayTypes.get(typeName)
  }

  getViewType(typeName: string): ViewType {
    return this.viewTypes.get(typeName)
  }

  getWidgetType(typeName: string): WidgetType {
    return this.widgetTypes.get(typeName)
  }

  getConnectionType(typeName: string): ConnectionType {
    return this.connectionTypes.get(typeName)
  }

  getRpcMethodType(methodName: string): RpcMethodType {
    return this.rpcMethods.get(methodName)
  }

  addRendererType(
    creationCallback: (pluginManager: PluginManager) => RendererType,
  ): this {
    return this.addElementType('renderer', creationCallback)
  }

  addAdapterType(
    creationCallback: (pluginManager: PluginManager) => AdapterType,
  ): this {
    return this.addElementType('adapter', creationCallback)
  }

  addTrackType(
    creationCallback: (pluginManager: PluginManager) => TrackType,
  ): this {
    // Goes through the already-created displays and registers the ones that
    // specify this track type
    const callback = () => {
      const track = creationCallback(this)
      ;(this.getElementTypesInGroup('display') as DisplayType[]).forEach(
        display => {
          if (
            display.trackType === track.name &&
            // track may have already added the displayType in its creationCallback
            !track.displayTypes.includes(display)
          ) {
            track.addDisplayType(display)
          }
        },
      )
      return track
    }
    return this.addElementType('track', callback)
  }

  addDisplayType(
    creationCallback: (pluginManager: PluginManager) => DisplayType,
  ): this {
    return this.addElementType('display', creationCallback)
  }

  addViewType(
    creationCallback: (pluginManager: PluginManager) => ViewType,
  ): this {
    const callback = () => {
      const newView = creationCallback(this)
      ;(this.getElementTypesInGroup('display') as DisplayType[]).forEach(
        display => {
          if (
            display.viewType === newView.name &&
            // view may have already added the displayType in its creationCallback
            !newView.displayTypes.includes(display)
          ) {
            newView.addDisplayType(display)
          }
        },
      )
      return newView
    }
    return this.addElementType('view', callback)
  }

  addWidgetType(
    creationCallback: (pluginManager: PluginManager) => WidgetType,
  ): this {
    return this.addElementType('widget', creationCallback)
  }

  addConnectionType(
    creationCallback: (pluginManager: PluginManager) => ConnectionType,
  ): this {
    return this.addElementType('connection', creationCallback)
  }

  addRpcMethod(
    creationCallback: (pluginManager: PluginManager) => RpcMethodType,
  ): this {
    return this.addElementType('rpc method', creationCallback)
  }
}
