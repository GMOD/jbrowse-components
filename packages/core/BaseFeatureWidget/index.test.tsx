import React from 'react'
import { ThemeProvider } from '@mui/material'
import { render } from '@testing-library/react'
import { types } from 'mobx-state-tree'

// locals
import { stateModelFactory } from '.'
import PluginManager from '../PluginManager'
import { ConfigurationSchema } from '../configuration'
import BaseFeatureDetails from './BaseFeatureDetail'
import { createJBrowseTheme } from '../ui'

test('open up a widget', async () => {
  const pluginManager = new PluginManager([])

  const Session = types.model({
    rpcManager: types.optional(types.frozen(), {}),
    configuration: ConfigurationSchema('test', {}),
    widget: stateModelFactory(pluginManager),
  })
  const model = Session.create(
    {
      widget: {
        type: 'BaseFeatureWidget',
      },
    },
    { pluginManager },
  )
  const { container, findByText } = render(
    <ThemeProvider theme={createJBrowseTheme()}>
      <BaseFeatureDetails model={model.widget} />
    </ThemeProvider>,
  )
  model.widget.setFeatureData({
    start: 2,
    end: 102,
    strand: 1,
    score: 37,
    refName: 'ctgA',
  })
  expect(await findByText('ctgA:3..102 (+)')).toBeTruthy()
  expect(container).toMatchSnapshot()
})
