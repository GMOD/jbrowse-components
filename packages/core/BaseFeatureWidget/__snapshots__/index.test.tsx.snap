// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`open up a widget 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiAccordion-root MuiAccordion-rounded Mui-expanded css-9g9pc9-MuiPaper-root-MuiAccordion-root"
    style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
  >
    <h3
      class="MuiAccordion-heading css-cy7rkm-MuiAccordion-heading"
    >
      <div
        aria-expanded="true"
        class="MuiButtonBase-root MuiAccordionSummary-root Mui-expanded css-t1q4kr-MuiButtonBase-root-MuiAccordionSummary-root"
        role="button"
        tabindex="0"
      >
        <div
          class="MuiAccordionSummary-content Mui-expanded css-rof4d7-MuiAccordionSummary-content"
        >
          <span
            class="MuiTypography-root MuiTypography-button css-l5w3u1-MuiTypography-root"
          />
        </div>
        <div
          class="MuiAccordionSummary-expandIconWrapper Mui-expanded css-1wqf3nl-MuiAccordionSummary-expandIconWrapper"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1qtxdat-MuiSvgIcon-root-icon"
            data-testid="ExpandMoreIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
            />
          </svg>
        </div>
      </div>
    </h3>
    <div
      class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-entered css-qr6njo-MuiCollapse-root"
      style="min-height: 0px;"
    >
      <div
        class="MuiCollapse-wrapper MuiCollapse-vertical css-1x6hinx-MuiCollapse-wrapper"
      >
        <div
          class="MuiCollapse-wrapperInner MuiCollapse-vertical css-1i4ywhz-MuiCollapse-wrapperInner"
        >
          <div
            class="MuiAccordion-region"
            role="region"
          >
            <div
              class="MuiAccordionDetails-root css-1eji0yd-MuiAccordionDetails-root-expansionPanelDetails"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-1xfjy4j-MuiTypography-root"
              >
                Core details
              </p>
              <div
                class="css-1m8nxnb-field"
              >
                <div
                  class="css-ify9vk-fieldName"
                >
                  Position
                </div>
                <div
                  class="css-xb19vs-fieldValue"
                >
                  ctgA:3..102 (+)
                </div>
              </div>
              <div
                class="css-1m8nxnb-field"
              >
                <div
                  class="css-ify9vk-fieldName"
                >
                  Length
                </div>
                <div
                  class="css-xb19vs-fieldValue"
                >
                  <span>
                    100
                  </span>
                </div>
              </div>
              <hr
                class="MuiDivider-root MuiDivider-fullWidth css-1jpc804-MuiDivider-root"
              />
              <p
                class="MuiTypography-root MuiTypography-body1 css-1xfjy4j-MuiTypography-root"
              >
                Attributes
              </p>
              <div
                class="css-1m8nxnb-field"
              >
                <div
                  class="css-ify9vk-fieldName"
                  style="width: 40px;"
                >
                  score
                </div>
                <div
                  class="css-xb19vs-fieldValue"
                >
                  <span>
                    37
                  </span>
                </div>
              </div>
              <div
                class="css-57ilie-container"
              >
                <div
                  class="MuiFormControl-root MuiFormControl-marginDense css-h2948o-MuiFormControl-root-formControl"
                >
                  <button
                    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-colorPrimary css-12qtcjg-MuiButtonBase-root-MuiButton-root"
                    tabindex="0"
                    type="button"
                  >
                    Show feature sequence
                  </button>
                </div>
                <button
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1d2ld9h-MuiButtonBase-root-MuiIconButton-root"
                  tabindex="0"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-14500qs-MuiSvgIcon-root"
                    data-testid="HelpIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m1 17h-2v-2h2zm2.07-7.75-.9.92C13.45 12.9 13 13.5 13 15h-2v-.5c0-1.1.45-2.1 1.17-2.83l1.24-1.26c.37-.36.59-.86.59-1.41 0-1.1-.9-2-2-2s-2 .9-2 2H8c0-2.21 1.79-4 4-4s4 1.79 4 4c0 .88-.36 1.68-.93 2.25"
                    />
                  </svg>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
